distribution/push_v2.go

```diff
@@ -291,86 +302,141 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
-	mountAttemptsRemaining := 3
-
-	// Attempt to find another repository in the same registry to mount the layer
-	// from to avoid an unnecessary upload.
-	// Note: metadata is stored from oldest to newest, so we iterate through this
-	// slice in reverse to maximize our chances of the blob still existing in the
-	// remote repository.
-	for i := len(v2Metadata) - 1; i >= 0 && mountAttemptsRemaining > 0; i-- {
-		mountFrom := v2Metadata[i]
 
-		sourceRepo, err := reference.ParseNamed(mountFrom.SourceRepository)
+	// Attempt to find another repository in the same registry to mount the layer from to avoid an unnecessary upload
+	candidates := getRepositoryMountCandidates(pd.repoInfo, pd.authConfigHash, maxRepositoryMountAttempts, v2Metadata)
+	for _, mountCandidate := range candidates {
+		logrus.Debugf("attempting to mount layer %s (%s) from %s", diffID, mountCandidate.Digest, mountCandidate.SourceRepository)
+		desc, mounted, lu, err := pd.attemptBlobMount(ctx, progressOutput, diffID, pd.authConfigHash, mountCandidate)
 		if err != nil {
+			logrus.Infof("failed to mount layer %s (%s) from %s: %v", diffID, mountCandidate.Digest, mountCandidate.SourceRepository, err)
 			continue
 		}
-		if pd.repoInfo.Hostname() != sourceRepo.Hostname() {
-			// don't mount blobs from another registry
-			continue
+		if mounted {
+			logrus.Debugf("layer %s (%s) mounted from %s", diffID, mountCandidate.Digest, mountCandidate.SourceRepository)
+			return desc, nil
 		}
+		// cancel previous upload
+		cancelLayerUpload(ctx, mountCandidate.Digest, layerUpload)
+		layerUpload = lu
+	}
 
+	if layerUpload == nil {
+		layerUpload, err = pd.repo.Blobs(ctx).Create(ctx)
+		if err != nil {
+			return distribution.Descriptor{}, retryOnError(err)
+		}
+	}
+	defer layerUpload.Close()
+
+	// upload the blob
+	desc, err := pd.uploadUsingSession(ctx, progressOutput, diffID, layerUpload)
+	if err != nil {
+		return desc, err
+	}
+
+	pd.pushState.Lock()
+	// If Commit succeeded, that's an indication that the remote registry speaks the v2 protocol.
+	pd.pushState.confirmedV2 = true
+	pd.pushState.remoteLayers[diffID] = desc
+	pd.pushState.Unlock()
+
+	return desc, nil
+}
+
+func (pd *v2PushDescriptor) SetRemoteDescriptor(descriptor distribution.Descriptor) {
+	pd.remoteDescriptor = descriptor
+}
+
+func (pd *v2PushDescriptor) Descriptor() distribution.Descriptor {
+	return pd.remoteDescriptor
+}
+
+// attemptBlobMount attempts to mount blob from the given source repository if given. If not given or
+// the mount fails, an upload will be initiated and layerUpload returned. If the mount succeeds, blob
+// descriptor will be returned with mounted == "true".
+func (pd *v2PushDescriptor) attemptBlobMount(
+	ctx context.Context,
+	progressOutput progress.Output,
+	diffID layer.DiffID,
+	authConfigHash string,
+	mountFrom metadata.V2Metadata,
+) (
+	desc distribution.Descriptor,
+	mounted bool,
+	layerUpload distribution.BlobWriter,
+	err error,
+) {
```

Comment:
Four return values feels excessive. Since this is only called in one place, it might be better to leave the code in-line instead of splitting it out into this function.